### PR TITLE
Debugging for `No applicable devices found` issue

### PR DESCRIPTION
**Pull request by iwasrobbed**
Hope this helps! I wasn't sure where to look when I first stumbled upon this error and it seemed like there were so many different symptoms causing it.

Addresses: https://github.com/fastlane/fastlane/issues/8737

### CODE REVIEW

**Comment by iwasrobbed** [comment]
I signed it!

File: docs/codesigning/common-issues.md [hunk 1]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
-## fastlane errors
+## _fastlane_ errors
 
 > Could not find a matching code signing identity for type 'X'
 
 There are no certificates available on the Apple Developer Portal. This could either mean someone revoked a certificate, or you don't have access to it. 
+
+# Debugging Steps for "No applicable devices found." error
+
+It's not immediately obvious what this error is or what it's caused by, so hopefully these simple steps will help you find the root cause faster than my experience with it.
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: We usually try to saying `I` or `My`

File: docs/codesigning/common-issues.md [hunk 2]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
 Make sure to have a valid code signing identity defined in your project targets. This might happen when you select `Don't Code Sign` as Code Signing Identity.
 
-## fastlane errors
+## _fastlane_ errors
 
 > Could not find a matching code signing identity for type 'X'
 
 There are no certificates available on the Apple Developer Portal. This could either mean someone revoked a certificate, or you don't have access to it. 
+
+# Debugging Steps for "No applicable devices found." error
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: lower case steps please

File: docs/codesigning/common-issues.md [hunk 3]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+
+It's not immediately obvious what this error is or what it's caused by, so hopefully these simple steps will help you find the root cause faster than my experience with it.
+
+### Not Just _fastlane_
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: lower case just

File: docs/codesigning/common-issues.md [hunk 4]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+
+This error happens to all projects built with `xcodebuild` / `xcrun`, so if you're trying to blame _fastlane_ here, take a deep breath and look elsewhere first. _fastlane_ might help alleviate it with some env var unsetting / setting or other tricks (see below), but this is an error from the new export APIs that Xcode uses.
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: Let's replace `so if you're trying to blame _fastlane_ here, take a deep breath and look elsewhere first` with `and it not caused by _fastlane_`

File: docs/codesigning/common-issues.md [hunk 5]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+
+### Legacy Export API
+
+As of Xcode 8.3, the deprecated / legacy export api (e.g. `use_legacy_build_api`) is removed so you're unfortunately stuck with using the latest, and often broken, export API.
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: Let's replace `is removed so you're unfortunately stuck with using the latest, and often broken, export API` with `is removed, so you have to use the latest Xcode export API, which unfortunately still has some issues.

File: docs/codesigning/common-issues.md [hunk 6]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+
+### Red Herring
+
+The error message, `"No applicable devices found."`, is actually just a red herring: it's not the real error. Something else is erroring out under the hood so you'll need to dig in to find out what.
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: What's a `red herring`, maybe let's use a phrase that's more commonly known for non-native speakers

File: docs/codesigning/common-issues.md [hunk 7]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+
+- [CFPropertyList](https://github.com/ckruse/CFPropertyList)
+- [sqlite3](https://github.com/sparklemotion/sqlite3-ruby)
+- [digest](https://rubygems.org/gems/digest)
+
+You can use `$ gem list` to show which gems are installed
+
+### Use `system` Ruby
+
+_fastlane_ will force you to use system Ruby version, by unsetting some env vars. This caused many of the old instances of this problem, but likely won't impact many in the future.
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: This has to be rephrased, as we don't force people to use system ruby

File: docs/codesigning/common-issues.md [hunk 8]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+
+It should be perfectly fine to use RVM and Xcode may even work nicely with newer Ruby versions, so just keep this option in mind for legacy builds.
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: What does that mean?

File: docs/codesigning/common-issues.md [hunk 9]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+$ rvm use system
+$ ruby -v
+ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]
+$ which ruby
+/usr/bin/ruby
+```
+
+### Check the Logs
+
+In my situation and in many that I saw on this repo and across the interwebs, this error was given during the bundling process. Luckily, that bundler creates a log file that you can access to see what error *actually* happened:
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: Let's try to avoid `my` and `I`

File: docs/codesigning/common-issues.md [hunk 10]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+>1.2.840.113635.100.1.61
+>1.2.840.113635.100.1.61
+>2017-03-30 21:47:14.702 xcodebuild[78957:9114437] [MT] IDEDistribution: Step failed: <IDEDistributionThinningStep: 0x7fc3218d0b60>: Error Domain=IDEDistributionErrorDomain Code=14 "No applicable devices found." UserInfo={NSLocalizedDescription=No applicable devices found.}
+error: exportArchive: No applicable devices found.
+
+If you look closely, you'll see the red herring error message and then just above it, there's a path to a log file (`*.xcdistributionlogs`). Opening that up, you'll see more detail. (Side note: you can also search for all logs via `$ sudo find /private/var -name "*.xcdistributionlogs"`)
+
+In my case, it was because my export was built without bitcode, but I had static lib frameworks that were built with it (inside `IDEDistribution.standard.log` within the `.xcdistributionlogs` file):
+
+>2017-03-31 16:49:06 +0000  Validating IPA structure...
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: This should probably be a code block instead

Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: The whole thing

File: docs/codesigning/common-issues.md [hunk 11]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+>2017-03-31 16:49:06 +0000  error: Failed to verify bitcode in opus.framework/opus:
+error: Bundle only contains bitcode-marker /var/folders/s1/0h7d0scx6xqdthft328yfk0m0000gn/T/IDEDistributionThinningStep.KRt/Payload/Viv.app/Frameworks/opus.framework/opus (armv7)
+
+I rebuilt that framework without bitcode and it worked like a charm! 😅
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: `I`

File: docs/codesigning/common-issues.md [hunk 12]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+
+### Other Causes
+
+Keep in mind, this list could be infinite since this `No applicable devices found` error just manifests itself from underlying errors. Your best bet is to dig into the logs and see which error is actually being thrown.
+
+- Using the wrong `export_method` or some other error in the generated `exportPlist` that gets generated
+- Root Apple WWDR certificate had expired; check the Keychain for expired certificates
+- Env vars not being unset, [relevant PR](https://github.com/fastlane-old/gym/pull/176)
+- Including bitcode sometimes caused it (can use `include_bitcode: false`) in your _gym_ command and check if all frameworks match your bitcode requirements
+- Provisioning profiles not matching capabilities and beta-reports-active flag on .entitlements file.
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: Isn't another error case that the gems you listed above are not installed on the system ruby? So running `gem install [gem_names]` for the system ruby should solve it, right? `rvm use system && gem install [gem_names]`

File: docs/codesigning/common-issues.md [hunk 13]
```diff
@@ -32,8 +32,121 @@ Your project defines a provisioning profile which doesn't exist on your local ma
...
+
+### Apple Tech Notes
+
+If all else fails, use these tech notes to troubleshoot app thinning and bitcode: 
+
+- https://developer.apple.com/library/content/technotes/tn2432/_index.html
+- https://stackoverflow.com/questions/34959767/whats-the-difference-between-fembed-bitcode-and-bitcode-generation-mode
+- https://forums.developer.apple.com/thread/70583
```
Comment (KrauseFx) [inline docs/codesigning/common-issues.md]: It would be awesome to also link to at least one radar here

**Comment by mfurtak** [comment]
This is going to be a really great resource - thanks for putting it together ❤️ 

Good comments, @KrauseFx :+1:

**Comment by iwasrobbed** [comment]
Thanks for the feedback @KrauseFx ; I'll clean this up tomorrow or early next week

**Comment by iwasrobbed** [comment]
Looks like the test build just needs a restart for a bundler issue; but otherwise let me know if all else looks alright @KrauseFx / @mfurtak 🙌

**Pull request merged**